File: backend/apps/games/dda.py
"""Pure dynamic-difficulty function shared conceptually with the web client."""

from dataclasses import asdict, dataclass, replace
from typing import Literal

Reason = Literal[
    "promote", "hold", "demote", "doctor_lock", "cap", "insufficient_data", "guest", "fatigue_hold"
]


@dataclass(frozen=True)
class SessionSummary:
    level: int
    accuracy: float
    meanReactionMs: float
    mistakes: int
    hintsUsed: int
    rounds: int
    completed: bool
    challengeMode: bool
    guestMode: bool
    fatigueFlagged: bool


@dataclass(frozen=True)
class DifficultyStateData:
    level: int
    window: list[dict]
    lockedByDoctor: bool
    capLevel: int | None
    minLevel: int
    maxLevel: int
    lockedByName: str = ""


@dataclass(frozen=True)
class DdaConfig:
    windowSize: int = 3
    promoteAccuracy: float = 0.85
    demoteAccuracy: float = 0.60
    reactionWorsenRatio: float = 1.25
    hintPenaltyPerRound: float = 0.5


@dataclass(frozen=True)
class DifficultyChangeData:
    fromLevel: int
    toLevel: int
    reasonCode: Reason
    explanation: str


@dataclass(frozen=True)
class DdaResult:
    state: DifficultyStateData
    change: DifficultyChangeData
    messageKey: str


def _result(
    state: DifficultyStateData,
    to_level: int,
    reason: Reason,
    explanation: str,
    message: str,
    *,
    clear: bool = False,
) -> DdaResult:
    return DdaResult(
        replace(state, level=to_level, window=[] if clear else state.window),
        DifficultyChangeData(state.level, to_level, reason, explanation),
        message,
    )


def next_difficulty(
    state: DifficultyStateData, session: SessionSummary, config: DdaConfig
) -> DdaResult:
    if session.guestMode:
        return _result(
            state,
            state.level,
            "guest",
            "Guest sessions do not affect difficulty.",
            "dda.thanks_for_playing",
        )
    window = [*state.window, asdict(session)][-config.windowSize :]
    updated = replace(state, window=window)
    if state.lockedByDoctor:
        name = state.lockedByName or "the care team"
        return _result(
            updated,
            state.level,
            "doctor_lock",
            f"Difficulty is locked by Dr. {name}.",
            "dda.same_next_time",
        )
    if len(window) < config.windowSize:
        return _result(
            updated,
            state.level,
            "insufficient_data",
            "More sessions are needed before adjusting difficulty.",
            "dda.same_next_time",
        )
    if session.fatigueFlagged or not session.completed:
        return _result(
            updated,
            state.level,
            "fatigue_hold",
            "Held because the latest session was cut short or a break was suggested.",
            "dda.thanks_for_playing",
        )
    all_low = all(item["accuracy"] < config.demoteAccuracy for item in window)
    earlier_mean = sum(item["meanReactionMs"] for item in window[:-1]) / len(window[:-1])
    slower = session.meanReactionMs >= config.reactionWorsenRatio * earlier_mean
    frequent = session.mistakes >= session.rounds / 2
    if all_low and (slower or frequent):
        target = max(state.minLevel, state.level - 1)
        why = "reaction time increased" if slower else "mistakes were frequent"
        return _result(
            updated,
            target,
            "demote",
            f"Accuracy stayed under 60% across 3 rounds while {why}.",
            "dda.easier_next_time",
            clear=target != state.level,
        )
    all_high = all(item["accuracy"] >= config.promoteAccuracy for item in window)
    low_hints = all(
        item["hintsUsed"] / max(item["rounds"], 1) < config.hintPenaltyPerRound for item in window
    )
    mean_rt = sum(item["meanReactionMs"] for item in window) / len(window)
    if all_high and low_hints and session.meanReactionMs <= 1.1 * mean_rt:
        ceiling = min(
            state.maxLevel, state.capLevel if state.capLevel is not None else state.maxLevel
        )
        target = min(ceiling, state.level + 1)
        if target == state.level and state.capLevel is not None:
            return _result(
                updated,
                target,
                "cap",
                f"Reached the doctor-set cap of level {state.capLevel}.",
                "dda.same_next_time",
                clear=True,
            )
        if target != state.level:
            return _result(
                updated,
                target,
                "promote",
                "Accuracy stayed at or above 85% with few hints.",
                "dda.harder_next_time",
                clear=True,
            )
    return _result(
        updated, state.level, "hold", "Performance within target range.", "dda.same_next_time"
    )

File: frontend/src/games/dda.ts
export interface SessionSummary {
  level: number;
  accuracy: number;
  meanReactionMs: number;
  mistakes: number;
  hintsUsed: number;
  rounds: number;
  completed: boolean;
  challengeMode: boolean;
  guestMode: boolean;
  fatigueFlagged: boolean;
}
export interface DifficultyStateData {
  level: number;
  window: SessionSummary[];
  lockedByDoctor: boolean;
  capLevel: number | null;
  minLevel: number;
  maxLevel: number;
  lockedByName?: string;
}
export interface DdaConfig {
  windowSize: number;
  promoteAccuracy: number;
  demoteAccuracy: number;
  reactionWorsenRatio: number;
  hintPenaltyPerRound: number;
}
export type ReasonCode =
  | "promote"
  | "hold"
  | "demote"
  | "doctor_lock"
  | "cap"
  | "insufficient_data"
  | "guest"
  | "fatigue_hold";
export interface DdaChange {
  fromLevel: number;
  toLevel: number;
  reasonCode: ReasonCode;
  explanation: string;
}
export interface DdaResult {
  state: DifficultyStateData;
  change: DdaChange;
  messageKey: string;
}
export const defaultDdaConfig: DdaConfig = {
  windowSize: 3,
  promoteAccuracy: 0.85,
  demoteAccuracy: 0.6,
  reactionWorsenRatio: 1.25,
  hintPenaltyPerRound: 0.5,
};

function result(
  state: DifficultyStateData,
  toLevel: number,
  reasonCode: ReasonCode,
  explanation: string,
  messageKey: string,
  clear = false,
): DdaResult {
  return {
    state: { ...state, level: toLevel, window: clear ? [] : state.window },
    change: { fromLevel: state.level, toLevel, reasonCode, explanation },
    messageKey,
  };
}

export function nextDifficulty(
  state: DifficultyStateData,
  session: SessionSummary,
  config: DdaConfig = defaultDdaConfig,
): DdaResult {
  if (session.guestMode)
    return result(
      state,
      state.level,
      "guest",
      "Guest sessions do not affect difficulty.",
      "dda.thanks_for_playing",
    );
  const updated = {
    ...state,
    window: [...state.window, session].slice(-config.windowSize),
  };
  if (state.lockedByDoctor)
    return result(
      updated,
      state.level,
      "doctor_lock",
      `Difficulty is locked by Dr. ${state.lockedByName ?? "the care team"}.`,
      "dda.same_next_time",
    );
  if (updated.window.length < config.windowSize)
    return result(
      updated,
      state.level,
      "insufficient_data",
      "More sessions are needed before adjusting difficulty.",
      "dda.same_next_time",
    );
  if (session.fatigueFlagged || !session.completed)
    return result(
      updated,
      state.level,
      "fatigue_hold",
      "Held because the latest session was cut short or a break was suggested.",
      "dda.thanks_for_playing",
    );
  const low = updated.window.every(
    (item) => item.accuracy < config.demoteAccuracy,
  );
  const earlier = updated.window.slice(0, -1);
  const earlierMean =
    earlier.reduce((sum, item) => sum + item.meanReactionMs, 0) /
    earlier.length;
  const slower =
    session.meanReactionMs >= config.reactionWorsenRatio * earlierMean;
  const frequent = session.mistakes >= session.rounds / 2;
  if (low && (slower || frequent)) {
    const target = Math.max(state.minLevel, state.level - 1);
    return result(
      updated,
      target,
      "demote",
      `Accuracy stayed under 60% across 3 rounds while ${slower ? "reaction time increased" : "mistakes were frequent"}.`,
      "dda.easier_next_time",
      target !== state.level,
    );
  }
  const high = updated.window.every(
    (item) => item.accuracy >= config.promoteAccuracy,
  );
  const fewHints = updated.window.every(
    (item) =>
      item.hintsUsed / Math.max(item.rounds, 1) < config.hintPenaltyPerRound,
  );
  const meanRt =
    updated.window.reduce((sum, item) => sum + item.meanReactionMs, 0) /
    updated.window.length;
  if (high && fewHints && session.meanReactionMs <= 1.1 * meanRt) {
    const ceiling = Math.min(state.maxLevel, state.capLevel ?? state.maxLevel);
    const target = Math.min(ceiling, state.level + 1);
    if (target === state.level && state.capLevel !== null)
      return result(
        updated,
        target,
        "cap",
        `Reached the doctor-set cap of level ${state.capLevel}.`,
        "dda.same_next_time",
        true,
      );
    if (target !== state.level)
      return result(
        updated,
        target,
        "promote",
        "Accuracy stayed at or above 85% with few hints.",
        "dda.harder_next_time",
        true,
      );
  }
  return result(
    updated,
    state.level,
    "hold",
    "Performance within target range.",
    "dda.same_next_time",
  );
}

export function sessionLevel(
  state: DifficultyStateData,
  challengeMode: boolean,
): number {
  return challengeMode
    ? Math.min(
        state.level + 1,
        state.capLevel ?? state.maxLevel,
        state.maxLevel,
      )
    : state.level;
}
